TiFBackendUtils/LocalSQL/MySQLDriver.ts

```diff
@@ -0,0 +1,172 @@
+    const conn = await this.conn
+    const [result] = await conn.execute<ResultSetHeader>(query, args)
+    if (isResultSetHeader(result)) {
+      return {
+        insertId: result.insertId.toString(),
+        affectedRows: result.affectedRows
+      }
+    } else {
+      throw new Error("Execution did not return a ResultSetHeader.")
+    }
+  }
+
+  /**
+   * Runs the given SQL query and returns a success result containing the insertId of the query.
+   */
+  executeResult (query: string, args: object | null = null) {
+    return promiseResult(
+      this.execute(query, args).then((executeResult) =>
+        success(executeResult)
+      )
+    )
+  }
+
+  /**
+   * Runs the given SQL query and returns a success result containing the result of the query.
+   */
+  queryResult<Value> (query: string, args: object | null = null) {
+    return promiseResult(
+      this.query<Value>(query, args).then((result) => success(result))
+    )
+  }
+
+  /**
+   * Performs an idempotent transaction and returns the result of the transaction wrapped in a {@link PromiseResult}.
+   */
+  transaction<SuccessValue, ErrorValue> (
+    querySQLTransaction: (tx: MySQLExecutableDriver) => AwaitableResult<SuccessValue, ErrorValue>
+  ) {
+    console.log("trying transaction")
+    return promiseResult((async () => {
+      const conn = await this.conn
+      try {
+        await conn.beginTransaction()
+        const result = await querySQLTransaction(this)
+        await conn.commit()
+        return result
+      } catch (error) {
+        console.log("Rollback called")
+        await conn.rollback()
+        throw error
+      }
+    })())
+  }
+
+  /**
+   * Loads a list of results from the database given a sql query and casts the result to `Value`
+   *
+   * @example
+   * ```ts
+   * type User = { id: string, ... }
+   *
+   * const results = await execute<User>(conn, "SELECT * FROM user");
+   * console.log(results[0].id) // ✅ Typesafe
+   * ```
+   */
+
+  async query<Value> (
+    query: string,
+    args: object | null
+  ): Promise<Value[]> {
+    // Use this.conn to execute the query and return the result rows
+    // This will be the only function to directly use the database library's execute method.
+    const conn = await this.conn
+    const [rows] = await conn.query(query, args)
+    if (Array.isArray(rows)) {
+      return castTypes(rows as RowDataPacket[]) as Value[]
+    } else {
+      throw new Error("Query did not return an array of rows.")
+    }
+  }
+
+  // ==================
+  // Generic Methods
+  // ==================
+
+  /**
+   * Returns the first result of a query in a typesafe manner.
+   *
+   * @example
+   * ```ts
+   * type User = { id: string, ... }
+   *
+   * const result? = await queryFirst<User>(conn, "SELECT * FROM user");
+   * console.log(result?.id) // ✅ Typesafe
+   * ```
+   */
+  queryFirstResult (query: string, args: object | null = null) {
```

Comment:
Why would we need a method for just the first result specifically? It seems like it's not very flexible